crates/memory/src/op_store/time_slice_hash_store.rs

```diff
@@ -0,0 +1,532 @@
+use kitsune2_api::{K2Error, K2Result};
+
+/// In-memory store for time slice hashes.
+///
+/// The inner store will look something like this:
+/// `[ Block_0_3, Single_3_ABC, Block_4_2, Single_6_DEF ]`
+///
+/// Consecutive sequences of empty hashes are compressed into a single block. A block
+/// contains at least 1 hash.
+///
+/// Gaps are not permitted, so the representation starts out as a single block that spans
+/// the entire range of possible slice ids. As hashes are inserted, the block is split around
+/// the inserted hash.
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone))]
+pub(super) struct TimeSliceHashStore {
+    pub(super) highest_stored_id: Option<u64>,
+    inner: Vec<SliceHash>,
```

Comment:
If we use a sparse BTreeMap here instead of a Vec, the inserting would be a lot simpler.

[And I don't think the empty block aggregation is too bad](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f5c3d005d7485b3f99f27b2a08ba80ad)

Reply:
Jeez... yes it would be wouldn't it. I over-complicated this a bunch by trying to actually represent the missing elements. 

What I was thinking was that we want to know which empty spaces have been touched and which haven't. In reality though, that information is not that useful. We want to be able to say to a peer "I have X time slices and this is the combined hash". If we haven't received any data in some time window since the last time slice that did have data then it would be valid for the other peer to check if we're missing something. If we aren't and it's genuinely been quiet, then looking up 0 hashes is cheap... So we can do without tracking it and this becomes really simple.

I just wish I'd thought of this, because as soon as I saw this comment I was wondering what on earth I was thinking :D There was a thought process, but having been through the exercise, this can be way simpler :)